### ISSUE TITLE
Algorithm error about target model sampling in Deepmind's version

### ISSUE THREAD

**Issue by sf-Liu** [feifeibear/LLMSpeculativeSampling#29]
Hello author, 

I appreciate your work in this repo contributing to the community. I just think some of the implementations are not align with the pseudo code descried in the paper. I would like to make a deep discussion with you. 

In the line 151 of [speculative_sampling_v2](https://github.com/feifeibear/LLMSpeculativeSampling/blob/main/sampling/speculative_sampling.py#L151) you generated the logits of target model `p` with the prediction `x` from draft model. Then compare  `p` and  `q ` at the lookahead part. Here I believe the math expression of `p` in your code is 
![eq1](https://latex.codecogs.com/svg.image?&space;q(x_{n&plus;1}|x_1,...,x_n,\widetilde{x}_1,...,\widetilde{x}_K),q(x_{n&plus;2}|x_1,...,x_n,\widetilde{x}_1,...,\widetilde{x}_K),...,q(x_{n&plus;K-1}|x_1,...,x_n,\widetilde{x}_1,...,\widetilde{x}_K))
Then in [line 167](https://github.com/feifeibear/LLMSpeculativeSampling/blob/1da363e9d2201663577aa2d90074853e5fda7812/sampling/speculative_sampling.py#L167), the comparison is actually 
![eq2](https://latex.codecogs.com/svg.image?min(1,\frac{q(x_{n&plus;1}|x_1,...,x_n,\widetilde{x}_1,...,\widetilde{x}_K)}{p(x_{n&plus;1}|x_1,...,x_n)}))
which is different with the equation in article,
![eq3](https://latex.codecogs.com/svg.image?min(1,\frac{q(x_{n&plus;1}|x_1,...,x_n)}{p(x_{n&plus;1}|x_1,...,x_n)}))
         
In another word, it's not correct to get `p` with all drafts `x` are given. Instead, should get `p` with adding `x` one by one. Here are toy codes to show that ![eq3](https://latex.codecogs.com/svg.image?logit(q(x_{n&plus;1}|x_1,...,x_n))\neq&space;logit(q(x_{n&plus;1}|x_1,...,x_n,\widetilde{x}_{n&plus;1})))

```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM
from sampling.utils import norm_logits, sample, max_fn
torch.manual_seed(123)
torch_device = 'cuda'

root_dir = "../model/"
approx_model_name = root_dir + "alpaca-2-1.3b"
tokenizer = AutoTokenizer.from_pretrained(approx_model_name, trust_remote_code=True)

Decoder().set_tokenizer(tokenizer)

print("finish loading models")

inp = "The quick brown fox"

max_len = 20
inp_ids = tokenizer.encode(inp, return_tensors='pt').to(torch_device)
model = AutoModelForCausalLM.from_pretrained(approx_model_name, 
                                                       torch_dtype=torch.float16,
                                                       device_map="auto",
                                                       trust_remote_code=True)
model.eval().to(torch_device)
p = model(inp_ids).logits
next_tok = sample(norm_logits(p[:, -1, :],  1, 0, 0))

inp_next_ids = torch.cat((inp_ids, next_tok), dim=1)
p_next = model(inp_next_ids).logits
same_next_tok = sample(norm_logits(p[:, -2, :], 1, 0, 0))

print("Token sampled from p(x|inp):", next_tok)
print("The same token sampled from p(x|inp, x):", same_next_tok)


>>>Token sampled from p(x|inp): tensor([[29871]], device='cuda:0')
>>>The same token sampled from p(x|inp, x): tensor([[4931]], device='cuda:0')
```
I will pull a request to modify this difference, please check the correctness and let me know if you have any question.

Thanks,

**Comment by feifeibear** [comment]
Hi, did you set the random seed to be the same for sampling? 
I am not sure about your claim. If we get p with adding x one by one, the Sps will work in a very low algorithm intensity?

**Comment by sf-Liu** [comment]
Hello @feifeibear. Yes, the random seed is the same for sampling.  

According to the paper, it's necessary to get the `p` with adding x one by one. Also according to that, the `p` is "computed in parallel". So I guess the method is feeding `p`s in batch [gamma, n+gamma, vocab_num] to the target model. And the latency is few increased comparing just processing a `p` with [1, n+gamma, vocab_num]. You don't have to concern about the low algorithm intensity. You could check my pull request for the implementation details, please.  

On the other hand, the latency of SP mainly depends on acceptance rates. ![eq](https://camo.githubusercontent.com/c00567b1094204201db2c01b0f33306fa27361fe3a9177fccff2f0b539ca0f63/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f6d696e28312c253543667261632537427128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e292537442537427028785f2537426e2b31253744253743785f312c2e2e2e2c785f6e2925374429) will bring higher acceptance rates than the one of ![image](https://camo.githubusercontent.com/633b1c43331d4e2bcf85ea13f4d1c6097b03f007a9258be000563cc34c473b97/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f6d696e28312c253543667261632537427128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e2c2535437769646574696c6465253742782537445f312c2e2e2e2c2535437769646574696c6465253742782537445f4b292537442537427028785f2537426e2b31253744253743785f312c2e2e2e2c785f6e2925374429). I think this is the key of the SP efficiency.

**Comment by sf-Liu** [comment]
Actually, regarding the discussion, the main conflict is "what p[ :, -n, :] stand for". We all know p[ :, -1, :] is the logits of  the predicted token, that is ![eq](https://latex.codecogs.com/svg.image?&space;q(x_{n&plus;1}|x_1,...,x_n)). But what is the meaning of p[ :, -2, :]? Does it stand for ![eq](https://latex.codecogs.com/svg.image?&space;q(x_n|x_1,...,x_n)) or ![eq](https://latex.codecogs.com/svg.image?&space;q(x_n|x_1,...,x_{n-1}))? Please let me know if you have any idea related to it.   

I did further experiment, which shows the acceptances rate of your version is just a bit less than mine. It makes me confused, too.

**Comment by feifeibear** [comment]
> Actually, regarding the discussion, the main conflict is "what p[ :, -n, :] stand for". We all know p[ :, -1, :] is the logits of the predicted token, that is ![eq](https://camo.githubusercontent.com/ae77efb868cf2443f1189ec0c58e2f8a80606ecaffe9330af656bfe93d8716de/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f2673706163653b7128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e29). But what is the meaning of p[ :, -2, :]? Does it stand for ![eq](https://camo.githubusercontent.com/7e4a1044968d3930020e02e406196b38dcb7ba83c4475823730c68de36263ddd/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f2673706163653b7128785f6e253743785f312c2e2e2e2c785f6e29) or ![eq](https://camo.githubusercontent.com/fba4a49c1eb824efbef1bbec56e5a88588e7b07aebb46529ce1f68a567500a7e/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f2673706163653b7128785f6e253743785f312c2e2e2e2c785f2537426e2d3125374429)? Please let me know if you have any idea related to it.
> 
> I did further experiment, which shows the acceptance rates of your version are just a bit less than mine. It makes me confused, too.

Thanks, I understood your updates and gave feedback on your wonderful PR. few questions.

1. why not change the speculative_sampling function?
2. I see you add batching for the v2 function? Can you give a test results to show the results?
3. can you paste a screenshot to prove the new code still functions? I want to make sure the code is still working well.

**Comment by sf-Liu** [comment]
Hello @feifeibear, thank you for your understanding. I have made revisions following your reviews in the PR. Here are replies to your concerns,

1. speculative_sampling function utilized KV Cache. Feeding batch tensors to the target model will cause additional errors. Currently, I don't plan to make further changes regarding speculative_sampling function.

    1.1. I really appreciate your additional implement with KV Cache. I believe it's not hard to make the KVCacheModel fit batch tensors. Just making some change on the sample function will work.

2. A test result about the new speculative_sampling function_v2 is shown below. It works well with Chinses-Llamma-2. You can take your own test as well.
![image](https://github.com/user-attachments/assets/6012d3b1-5841-445c-ba2e-d794b6d78ef2)

**Comment by haiduo** [comment]
> Hello author,
> 
> I appreciate your work in this repo contributing to the community. I just think some of the implementations are not align with the pseudo code descried in the paper. I would like to make a deep discussion with you.
> 
> In the line 151 of [speculative_sampling_v2](https://github.com/feifeibear/LLMSpeculativeSampling/blob/main/sampling/speculative_sampling.py#L151) you generated the logits of target model `p` with the prediction `x` from draft model. Then compare `p` and `q ` at the lookahead part. Here I believe the math expression of `p` in your code is ![eq1](https://camo.githubusercontent.com/faa14541a554ed374f39862889717755631f80fc1d5218a6babec20cc2437e6d/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f2673706163653b7128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e2c2535437769646574696c6465253742782537445f312c2e2e2e2c2535437769646574696c6465253742782537445f4b292c7128785f2537426e2b32253744253743785f312c2e2e2e2c785f6e2c2535437769646574696c6465253742782537445f312c2e2e2e2c2535437769646574696c6465253742782537445f4b292c2e2e2e2c7128785f2537426e2b4b2d31253744253743785f312c2e2e2e2c785f6e2c2535437769646574696c6465253742782537445f312c2e2e2e2c2535437769646574696c6465253742782537445f4b29) Then in [line 167](https://github.com/feifeibear/LLMSpeculativeSampling/blob/1da363e9d2201663577aa2d90074853e5fda7812/sampling/speculative_sampling.py#L167), the comparison is actually ![eq2](https://camo.githubusercontent.com/633b1c43331d4e2bcf85ea13f4d1c6097b03f007a9258be000563cc34c473b97/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f6d696e28312c253543667261632537427128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e2c2535437769646574696c6465253742782537445f312c2e2e2e2c2535437769646574696c6465253742782537445f4b292537442537427028785f2537426e2b31253744253743785f312c2e2e2e2c785f6e2925374429) which is different with the equation in article, ![eq3](https://camo.githubusercontent.com/c00567b1094204201db2c01b0f33306fa27361fe3a9177fccff2f0b539ca0f63/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f6d696e28312c253543667261632537427128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e292537442537427028785f2537426e2b31253744253743785f312c2e2e2e2c785f6e2925374429)
> 
> In another word, it's not correct to get `p` with all drafts `x` are given. Instead, should get `p` with adding `x` one by one. Here are toy codes to show that ![eq3](https://camo.githubusercontent.com/4d17ab454ecf03100a8e92cb10a469f24a59a478a0394ae6b45f2f3c5762c13f/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f6c6f676974287128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e29292535436e65712673706163653b6c6f676974287128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e2c2535437769646574696c6465253742782537445f2537426e2b312537442929)
> 
> ```
> import torch
> from transformers import AutoTokenizer, AutoModelForCausalLM
> from sampling.utils import norm_logits, sample, max_fn
> torch.manual_seed(123)
> torch_device = 'cuda'
> 
> root_dir = "../model/"
> approx_model_name = root_dir + "alpaca-2-1.3b"
> tokenizer = AutoTokenizer.from_pretrained(approx_model_name, trust_remote_code=True)
> 
> Decoder().set_tokenizer(tokenizer)
> 
> print("finish loading models")
> 
> inp = "The quick brown fox"
> 
> max_len = 20
> inp_ids = tokenizer.encode(inp, return_tensors='pt').to(torch_device)
> model = AutoModelForCausalLM.from_pretrained(approx_model_name, 
>                                                        torch_dtype=torch.float16,
>                                                        device_map="auto",
>                                                        trust_remote_code=True)
> model.eval().to(torch_device)
> p = model(inp_ids).logits
> next_tok = sample(norm_logits(p[:, -1, :],  1, 0, 0))
> 
> inp_next_ids = torch.cat((inp_ids, next_tok), dim=1)
> p_next = model(inp_next_ids).logits
> same_next_tok = sample(norm_logits(p[:, -2, :], 1, 0, 0))
> 
> print("Token sampled from p(x|inp):", next_tok)
> print("The same token sampled from p(x|inp, x):", same_next_tok)
> 
> 
> >>>Token sampled from p(x|inp): tensor([[29871]], device='cuda:0')
> >>>The same token sampled from p(x|inp, x): tensor([[4931]], device='cuda:0')
> ```
> 
> I will pull a request to modify this difference, please check the correctness and let me know if you have any question.
> 
> Thanks,

Hi @sf-Liu , thanks for your good comments. But I think you may have misunderstood the process of generating tokens by the target model. In fact, the parallelism of the target model comes from the mechanism of mask attention, so the formula you derived may be an error, that is, the so-called line 151 of speculative_sampling_v2 is as follows:

$q(x_{n+1}|x_1, \ldots, x_n)$  
$q(x_{n+2}|x_1,\ldots, x_n,\tilde{x}_1) $  
$\ldots $  

$q(x_{n+K-1}|x_1,\ldots, x_n,\tilde{x}_1,\ldots,\tilde{x}\_{K-2}\) $

In addition, p(x) and q(x) represent the logits ​​of the target model and draft model respectively, while code and paper are opposite.

**Comment by feifeibear** [comment]
Thanks, @haiduo , Appreciate for helping me review these comments.

**Comment by sf-Liu** [comment]
> Actually, regarding the discussion, the main conflict is "what p[ :, -n, :] stand for". We all know p[ :, -1, :] is the logits of the predicted token, that is ![eq](https://camo.githubusercontent.com/ae77efb868cf2443f1189ec0c58e2f8a80606ecaffe9330af656bfe93d8716de/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f2673706163653b7128785f2537426e2b31253744253743785f312c2e2e2e2c785f6e29). But what is the meaning of p[ :, -2, :]? Does it stand for ![eq](https://camo.githubusercontent.com/7e4a1044968d3930020e02e406196b38dcb7ba83c4475823730c68de36263ddd/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f2673706163653b7128785f6e253743785f312c2e2e2e2c785f6e29) or ![eq](https://camo.githubusercontent.com/fba4a49c1eb824efbef1bbec56e5a88588e7b07aebb46529ce1f68a567500a7e/68747470733a2f2f6c617465782e636f6465636f67732e636f6d2f7376672e696d6167653f2673706163653b7128785f6e253743785f312c2e2e2e2c785f2537426e2d3125374429)? Please let me know if you have any idea related to it.

Hello @haiduo, I'm glad more people are joining in our discussion. I checked the source code in `transformers`. Its method is the same as @feifeibear 's, which is convincing proof. 

I'm still confused about the meaning of such a method. Because the input to the target model is **all** the candidates generated by the draft model. (I checked the attention mask in `transformers` source codes and it's all ones.) So I suppose the conditions of the probabilities are always $x_1,..., x_n,  \widetilde x_1,..., \widetilde x_{K-1}$. Then it comes to the question I mentioned above. Can you make any clearification to help with my confusion? Perhaps it indeed makes the target model calculates $p(x_{n+1})$ and $p(x_{n+2})$ with different conditions. Did I miss anything?

I think the auther set the `p` `q` notations as them in `speculative_sampling` function and these two are opposite. It's not a big issue as long as we indicate the target/draft model clearly when we mention them.

**Comment by haiduo** [comment]
Hi @sf-Liu , first, you find that all the values ​​of the attention mask are equal to 1, which is true, but in fact, the [torch.nn.functional.scaled_dot_product_attention](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L610) of transformers replaces 'attn_mask=causal_mask' with 'is_causal', that is, if the value of the attention mask is not specially processed, the default is none, but is_causal is True. I guess this implementation is similar to making a lower triangular mask for attn_mask, see [hard-code](https://github.com/pytorch/pytorch/issues/108108) for details. The code implemented by the @feifeibear  is similar to that implemented by the transformer library, but both call the scaled_dot_product_attention function by default. Because this function is encapsulated, I don’t know the specific implementation. But in fact, the value of the  attention mask should be processed as follows: 
![attention mask](https://github.com/user-attachments/assets/de4b06df-a346-49fd-99bd-cc7b8c9dea4c)

I hope the above helps you.
In addition, provide a reasonable display and processing implementation of the attention mask:
```
 # Copied from transformers.models.bart.modeling_bart.BartDecoder._prepare_decoder_attention_mask
def _prepare_decoder_attention_mask(self, attention_mask, input_shape, inputs_embeds, past_key_values_length):
      # create causal mask
      # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
      combined_attention_mask = None
      if input_shape[-1] > 1:
          combined_attention_mask = _make_causal_mask(
              input_shape,
              # inputs_embeds.dtype,
              torch.float32,  # [MODIFIED] force to cast to float32
              device=inputs_embeds.device,
              past_key_values_length=past_key_values_length,
          )

      if attention_mask is not None:
          # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
          expanded_attn_mask = _expand_mask(attention_mask, inputs_embeds.dtype, tgt_len=input_shape[-1]).to(inputs_embeds.device)
          combined_attention_mask = (expanded_attn_mask if combined_attention_mask is None else expanded_attn_mask + combined_attention_mask)

      if hasattr(self, "tree_mask") and self.tree_mask is not None:
          tree_mask = self.tree_mask
          tree_len = tree_mask.size(-1)
          combined_attention_mask[:, :, -tree_len:, -tree_len:][tree_mask==0] = combined_attention_mask.min()

      return combined_attention_mask

if attention_mask is None:
        attention_mask = torch.ones((batch_size, seq_length_with_past), dtype=torch.bool, device=hidden_states.device,)
attention_mask = self._prepare_decoder_attention_mask(attention_mask, (batch_size, seq_length), hidden_states, past_key_values_length,)
```

**Comment by sf-Liu** [comment]
Cool @haiduo. It's a really nice explanation.
